### ISSUE TITLE
Succeeded jobs counted into parallelism limit

### ISSUE THREAD

**Issue by marko-ark** [argoproj/argo-workflows#12541]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issue exists when I tested with `:latest`
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

We have set our Argo Workflows with `parallelism: 2048` and `- --qps=2048`. We have no `namespaceParallelism`. We are also running 5 Workflow Controllers. The system runs up to 3000 workflows during the day in 2 different namespaces. However, we have noticed that we end up having a lot of Pending workflows with the message: `Workflow processing has been postponed because too many workflows are already running` while there are 0 `Running` workflows at that very moment, and 2000+ workflows marked as `Succeeded`. Our default TTL for workflows is 24h but we did add some in the last week that have a longer TTL - 3-4 days for some jobs.

The moment I manually delete all the `Succeeded` workflows using `kubectl delete wf -n argo -l workflows.argoproj.io/phase=Succeeded` the workflows get unclogged and everything runs as normal. While it is not a problem to lower the TTL for successful jobs I am wondering if this is intended behaviour or if we are missing something in our configuration. I have read the whole default `values.yaml` in the Helm chart for Argo Worfklows but I didn't see anything pointing to this particular issue.

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
N/A
```

### Logs from the workflow controller

```text
N/A
```

### Logs from in your workflow's wait container

```text
N/A
```

**Comment by agilgur5** [comment]
Completed Workflows should [be removed](https://github.com/argoproj/argo-workflows/blob/1ab7cd2071c6bfc335a89340d736138b20caf421/workflow/controller/controller.go#L794) from the parallelism count... Did this start happening after a recent update? (although this feature hasn't gotten much change in years, it's possible something affected it)

Also do these Workflows have the "Completed" label as well?

**Comment by tooptoop4** [comment]
@marko-ark what version didn't have this issue?

**Comment by tzahush** [comment]
I've just experienced the same problem when upgrading from helm `0.39.7` to helm `0.40.7`

**Comment by agilgur5** [comment]
Helm chart versioning is different from the underlying app (literally `version` vs `appVersion`), can be overridden as well, and Helm is not used by all users. Please use the Argo Workflows version in this repo.

> helm `0.39.7` to helm `0.40.7`

That would be from Argo [v3.5.2](https://github.com/argoproj/argo-helm/blob/argo-workflows-0.39.7/charts/argo-workflows/Chart.yaml#L2) to Argo [v3.5.4](https://github.com/argoproj/argo-helm/blob/argo-workflows-0.40.7/charts/argo-workflows/Chart.yaml#L2)

> Version '3.5.2' worked just fine, however, the upgrade to `3.5.3` is when it started.

That would match up with this report, although as OP said, it could just be coincidence. Let's investigate the changelog to see if anything relevant changed in those versions

**Comment by agilgur5** [comment]
I was suspecting #12198 as it had a small change to the parallelism logic, and indeed that was released as part of 3.5.3.

This may require adding a `throttler.Remove` in [the new tracking code](https://github.com/argoproj/argo-workflows/blame/1dbc856e51967feb58066a4087a8679b08b87be3/workflow/controller/controller.go#L958) and/or moving [this `defer`'d `throttler.Remove`](https://github.com/argoproj/argo-workflows/blame/1dbc856e51967feb58066a4087a8679b08b87be3/workflow/controller/controller.go#L790) up before the `RecentlyCompleted` [check](https://github.com/argoproj/argo-workflows/blame/1dbc856e51967feb58066a4087a8679b08b87be3/workflow/controller/controller.go#L771). Doing both actually might make sense since it should be idempotent and then we cover the bases.

cc @Joibel does that look like it could cause a possible regression to you?

**Comment by Joibel** [comment]
@agilgur5 I agree, I'll investigate and do a fix for this.

I don't think blindly moving the deferred Remove upwards before the attempted Admit would be correct, but somehow we do need a Remove in the RecentlyCompleted=true code path.

**Comment by Joibel** [comment]
I have been unable to reproduce this in a simple setup with the current tip of main or on v3.5.3.

I have set `parallelism` to 3, and with or without `namespaceParallelism` I get only 3 running workflows, and  
I'm leaving my workflows in the cluster. I just run a simple workflow
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: thing-
spec:
  entrypoint: thing
  templates:
  - name: thing
    container:
      image: alpine:3.6
      command: [sh, -c]
      args: ["sleep 60"]
```

Workflows are correctly running 3 at a time, waiting, and automatically starting as soon as the previous one finishes as expected.

Removing all retention policies from the controller didn't change anything, it still works as hoped.

I backported #12198 to workflows 3.4 for a large customer who're also running it with `parallelism` set, and they don't see this problem. I'm not saying #12198 isn't the cause of this, but I'd like a way to reproduce it before speculating a fix.

@marko-ark @tzahush, any chance of a copy of your workflow controller configmap (redacted as necessary)?

**Comment by agilgur5** [comment]
> @agilgur5 I agree, I'll investigate and do a fix for this.

Thanks for taking a look @Joibel !

> I backported #12198 to workflows 3.4 for a large customer who're also running it with `parallelism` set, and they don't see this problem. I'm not saying #12198 isn't the cause of this, but I'd like a way to reproduce it before speculating a fix.

Yea I wasn't sure if it was (that's why I pinged you 😅 ), but it's the only recent change in the `parallelism` logic. It _could_ be indirectly caused by something else though and/or not be a regression at all.
A very busy controller that has maxed out resources is the only other thing that comes to mind though (i.e. the `throttler.Remove` is heavily backed up).

> I have continued debugging things and I have noticed that some of the `Succeeded` workflows have `workflows.argoproj.io/completed=false` and that is due to sub-steps being in `Pending`state.

Well that sounds like a potential root cause -- the `reconciliationNeeded` function checks [this exact label](https://github.com/argoproj/argo-workflows/blob/baef4856ff2603c76dbe277c825eaa3f9788fc91/workflow/controller/controller.go#L836). I'm wondering if there was a race condition introduced recently into [the operator logic](https://github.com/argoproj/argo-workflows/blob/baef4856ff2603c76dbe277c825eaa3f9788fc91/workflow/controller/operator.go#L2326) that keeps the `phase` and `completed` label in sync somewhere 🤔

**Comment by agilgur5** [comment]
> I'm wondering if there was a race condition introduced recently into [the operator logic](https://github.com/argoproj/argo-workflows/blob/baef4856ff2603c76dbe277c825eaa3f9788fc91/workflow/controller/operator.go#L2326) that keeps the `phase` and `completed` label in sync somewhere 🤔

Might be another red herring, but #12284 made a tiny change to some of this logic and was released in v3.5.3 as well. Off hand the logic doesn't look incorrect to me, but it's possible there was some unsynchronized logic elsewhere that is now problematic with that code.

**Comment by Joibel** [comment]
The workflows in `workflows.argoproj.io/completed=false` seem potentially attributable to #12284. I'm still not ruling out #12198 though.

@marko-ark - could you get the controller logs filtered by a workflow that's in that state? The label being wrong would cause the issue you're seeing with parallelism.

@tzahush - can you tell us if you're seeing the same thing as @marko-ark?

**Comment by Joibel** [comment]
@Garett-MacGowan, Task Results Completion Status is new in #12402, and from a naive look at the code what @marko-ark is describing shouldn't be happening or possible. Are you aware of this?

It seems likely it's the cause of his issues here.

**Comment by Garett-MacGowan** [comment]
@juliev0 tagging for visibility.

**Comment by Garett-MacGowan** [comment]
@marko-ark Are you using a data template?

**Comment by juliev0** [comment]
> @marko-ark Are you using a data template?

Or could it also be a "resource" template @Garett-MacGowan (per [this](https://github.com/argoproj/argo-workflows/blob/main/test/e2e/testdata/resource-templates/outputs.yaml#L9) example)? It looks like [this](https://github.com/argoproj/argo-workflows/pull/12544) PR is related to both types. 

I see that in [#12402](https://github.com/argoproj/argo-workflows/pull/12402) setting `LabelKeyCompleted=true` is gated on the TaskResults being completed [here](https://github.com/argoproj/argo-workflows/pull/12402/files#diff-f321d4af83fcf8311dc80c0d50c59ac4c240f761206e7bb652709870eb9feb43R2309), so if those types of Resources never had TaskResults indicated as completed, I suppose it would prevent reaching that, huh?

```
if woc.checkReconciliationComplete() && !woc.hasDaemonNodes() {
			....
			woc.wf.ObjectMeta.Labels[common.LabelKeyCompleted] = "true"
```

**Comment by Garett-MacGowan** [comment]
> @Garett-MacGowan No, we are not.

Are you using a resource template? Could you post a redacted version of one of the workflows that has `workflows.argoproj.io/completed=false`. E.g. `etl-connector-data` 🙏

**Comment by Garett-MacGowan** [comment]
@marko-ark hmm. Are you able to identify where the following node IDs are coming from?
```
etl-connector-data-wphhw-2752339363
etl-connector-data-wphhw-878891159
```

**Comment by tzahush** [comment]
> The workflows in `workflows.argoproj.io/completed=false` seem potentially attributable to #12284. I'm still not ruling out #12198 though.
> 
> @marko-ark - could you get the controller logs filtered by a workflow that's in that state? The label being wrong would cause the issue you're seeing with parallelism.
> 
> @tzahush - can you tell us if you're seeing the same thing as @marko-ark?

Yes seems like it is the same case where althought the workflows are on status 'Succeed' they still have the label:
`"workflows.argoproj.io/completed=false"`

**Comment by agilgur5** [comment]
Ok so I think this is fixed by #12537 (with #12544 being a related fix but not specific to this issue), the root cause is #12402, and the rest of the issues are red herrings.

Anyone else please correct me if I'm wrong there.

**Comment by Garett-MacGowan** [comment]
> Ok so I think this is fixed by #12537 (with #12544 being a related fix but not specific to this issue), the root cause is #12402, and the rest of the issues are red herrings.
> 
> Anyone else please correct me if I'm wrong there.

Based on the available information, you're probably right.

**Comment by juliev0** [comment]
> Ok so I think this is fixed by #12537 (with #12544 being a related fix but not specific to this issue), the root cause is #12402, and the rest of the issues are red herrings.
> 
> Anyone else please correct me if I'm wrong there.

Are you thinking he would’ve gotten the same error from #12357 when evaluating step output?

**Comment by Garett-MacGowan** [comment]
> > Ok so I think this is fixed by #12537 (with #12544 being a related fix but not specific to this issue), the root cause is #12402, and the rest of the issues are red herrings.
> > Anyone else please correct me if I'm wrong there.
> 
> Are you thinking he would’ve gotten the same error from #12357 when evaluating step output? I believe that would result in a Failed Workflow, wouldn’t it?
> 
> Or otherwise can you help connect the dots? (Sorry, I’m on my phone right now so not looking at the code)

You mean #12537 ?

**Comment by juliev0** [comment]
> 
> You mean #12537 ?

Hah, yeah, sorry, I mistyped.

**Comment by agilgur5** [comment]
> Are you thinking he would’ve gotten the same error from #12357 when evaluating step output?

not necessarily, but the steps are marked as `false` [above](https://github.com/argoproj/argo-workflows/issues/12541#issuecomment-1907729727) (controller log is also attached in that comment, I don't see the error), so I imagine #12537 will impact / resolve that, as well as resolve the desync of the `completed` label and the `phase` label

If not, do you suspect there is another bug? If so, feel free to reopen, you know this code better than me for sure, and I did not read it in-depth (more of a cursory look)

**Comment by juliev0** [comment]
@Garett-MacGowan if you're able to explain how #12537 would resolve this, we can keep it closed. If not, I'd say we better reopen this.

**Comment by Garett-MacGowan** [comment]
> @Garett-MacGowan if you're able to explain how #12537 would resolve this, we can keep it closed. If not, I'd say we better reopen this.

We don't have full info on this ticket, so it might be good to keep it open. Also I haven't traced the code regarding parallelism, so I'm not 100% confident.

However, if `workflows.argoproj.io/completed=false` workflows contribute to the parallelism count, then it makes sense that this issue may be fixed now.

**Comment by agilgur5** [comment]
> However, if `workflows.argoproj.io/completed=false` workflows contribute to the parallelism count, then it makes sense that this issue may be fixed now.

Yes, they do, that's what we discovered was the issue earlier.

@marko-ark @tzahush could either of you test `latest` to see if this still occurs?

**Comment by Joibel** [comment]
@marko-ark, with the fix for #12537 you shouldn't be seeing `Succeeded` but `completed=false` workflows if I've understood the fixes correctly. 

Are the `Task Results Completion Status` still showing some tasks as `false` in the `completed=false` workflows?

**Comment by Garett-MacGowan** [comment]
> > However, if `workflows.argoproj.io/completed=false` workflows contribute to the parallelism count, then it makes sense that this issue may be fixed now.
> 
> Yes, they do, that's what we discovered was the commonality earlier in the thread.
> 
> @marko-ark @tzahush could either of you test `latest` to see if this still occurs?

@agilgur5 

Has `latest` (3.5.4) integrated these changes? I don’t think so, but correct me if I'm wrong. If so, @marko-ark are you using imagePullPolicy: always? You could be using a stale image.

**Comment by Garett-MacGowan** [comment]
> @agilgur5 Sadly no, this doesn't fix things for us. We have ran twice the whole pipeline today. The moment the number of workflows with `completed=false` goes above the parallelism limit we start seeing `Pending` workflows and things get stuck. However, the moment I manually delete all `Succeeded` but `completed=failed` workflows, or at least 50% of our parallelism limit (2048) things get unstuck right away and the number of `Pending` workflows goes to 0 within a minute.
> 
> @Garett-MacGowan Sorry, I completely missed your question. Those two IDs are workflows step IDs for each of the pods. The pods are marked as successful but the steps are not.

I will look into this today. I'm traveling, so it will be a bit later.

**Comment by agilgur5** [comment]
> Has `latest` (3.5.4)

The `latest` image is the the tip of the `main` branch, not the newest release

> If so, @marko-ark are you using imagePullPolicy: always? You could be using a stale image.

That is possible regardless though, yes, good call

**Comment by Garett-MacGowan** [comment]
@marko-ark 

>  Those two IDs are workflows step IDs for each of the pods. The pods are marked as successful but the steps are not.

Which steps are these refering to? Is the example workflow above the same one that produced those two IDs?

If you could test this on `latest` with `imagePullPolicy: always` & let me know if it is still failing, that would be ideal. If it is still failing, could you please create a minimal example workflow using public images (e.g. `argosay:v2`) that reproduces it?

If it is still failing, and you provide the above, we can re-open this ticket, and I will work towards a solution.

**Comment by Garett-MacGowan** [comment]
> @Garett-MacGowan
> 
> > Which steps are these refering to? Is the example workflow above the same one that produced those two IDs?
> 
> ```
> Task Results Completion Status:
>     etl-connector-data-wphhw-0-prepare-triggers-878891159:   true
>     etl-connector-data-wphhw-1-fan-out-triggers-2752339363:  true
>     etl-connector-data-wphhw-2752339363:                     false
>     etl-connector-data-wphhw-878891159:                      false
> ```
> 
> The first two are pods (true) and the last two are steps (false) from what we see in the UI.
> 
> You were correct, we didn't have proper `imagePullPolicy` in place so we got a stale image. Now we have `latest` as the tag and `imagePullPolicy:always`. We re-ran all our workflows multiple times and the moment the number of those that have `completed=false` goes above 2048 (our parallelism limit) pending starts. I have also noticed that this doesn't apply only to `Succeeded` but to `Failed` ones too. As long as the lable is `workflows.argoproj.io/completed=false` pending will happen.
> 
> I will now setup a minimal workflows as per your suggestion and test that. I will get back to you a bit later with the results. Is there anything specific you want that workflow to do?

Just something simple that I can add to the tests suite. Any workflow that gets stuck with`workflows.argoproj.io/completed=false`

**Comment by Garett-MacGowan** [comment]
> @Garett-MacGowan We have managed to find the source of the problem. TL;DR: service accounts lacked privilege to patch `workflowtaskresults`.
> 
> Since we had thousands of workflows it was rather hard to easily find a pattern. In the end we figured out that workflows that use the service account created by the Helm chart work just fine in every way, no `completed=false` and deletion worked OK. We have two other sets of workflows that use separate service accounts. After running things in there, but just a single workflow, we figured out that it is only in those namespace that the problem appears. Service Account came to mind right away. I checked all 3 service accounts and noticed that the two new SA missed `-patch` permission for `workflowtaskresults`. I will assume that they need this to update the label. Once I've added the correct permissions not a single successful job had `completed=false` and this morning our setup was cleaned up as expected.
> 
> We will keep the current `latest` version in place and then upgrade to the next release proper once it's available.
> 
> Thank you all for your time and help provided!

Great! I'll think on this and see if anything needs to be done docs wise or otherwise!

**Comment by Garett-MacGowan** [comment]
For anyone that comes across this issue in the future, this requirement is already mentioned in the docs [here](https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/)
